### PR TITLE
Added backup pruner cron

### PR DESCRIPTION
**Pull request by bdero**
This adds a script that runs all of the backup pruners and logs a report using GNU Parallel, as well as a cron job that runs this script twice daily.
If the pruning fails for any tarsnap keys, the logs for all of the failed pruner scripts will be collected and sent in one email.

@smarnach @itsjeyd

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,9 +6,7 @@ This server is used only to delete old backups.
 How to deploy
 -------------
 
-1. Set IP in hosts 
-2. Download vault password 
-3. `deploy-all.yaml -u admin --vault-password-file .vault-pass`
+`ansible-playbook deploy/deploy-all.yml -u ubuntu --extra-vars @private-extra-vars.yml -i hosts --private-key ~/.ssh/id_rsa_backup_pruner`
```
Comment (smarnach) [inline README.md]: You could use some placeholder instead of including the path of the private key on your system.

File: backup-pruner.yml [hunk 2]
```diff
@@ -46,3 +51,18 @@
...
+    src: "backup-pruner.sh"
+    dest: "{{ BACKUP_PRUNER_SCRIPT }}"
+    owner: "root"
+    group: "root"
+    mode: "700"
+
+- name: add backup pruner cron
+  cron:
+    name: "backup pruner"
+    minute: 30
```
Comment (smarnach) [inline backup-pruner.yml]: We are running the backups at 10 past the hour, so we might want to move this to 40 to maximize the distance (unless pruning typically takes longer than creating the backup in the first place).

File: templates/backup-pruner.sh [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+# Check {{ JOB_LOG }} for a summary of the most recent pruner run.
+set -e
+
+jobs='{% for item in TARSNAPPER_JOBS %}
+{{ item.name }}{% endfor %}'
+
+# Buffer all of the pruner output; if the pruner fails, forward along the
+# output to include in a cron email.
+wrapper='output=$(/usr/local/sbin/tarsnap-{}.sh expire 2>&1) || echo "$output"'
+echo "$jobs" | tail -n +2 | parallel --no-notice -j0 --joblog {{ JOB_LOG }} $wrapper
```
Comment (smarnach) [inline templates/backup-pruner.sh]: This is slightly hard to parse for human readers.  Why is the output first stored in a shell variable?  What detects whether the output should be mailed in the cron email?

**Review by smarnach — CHANGES_REQUESTED** [review]
@bdero I've had a few minor comments about your changes, but they look mostly fine.  However, the existing design of the backup pruner looks somewhat weird to me.  I don't understand why it's necessary to create one script per pruned backup.  My preferred approach would be to just have a single script for everything (including the cron job), and to have the configuration in a separate file, so the script is just copied and does not change based on what backups are pruned.  (Changing this is of course out of scope for the current story.)

File: templates/backup-pruner.sh (hunk 3 above)
Comment (bdero) [reply templates/backup-pruner.sh]: Parallel does write out a nice little log containing the exit code and total running time of each of the subprocesses, but I found the tarsnapper output is usually pretty good at saying exactly what went wrong, so I wanted to make sure the cron emails we receive contain the output for any failing tarsnapper runs so that we'll know immediately how serious the problem is.

To accomplish this I just wrapped the task in a form like this:
```
output=$(task_to_execute 2>&1) || echo "$output"
```
This just buffers the output into `output` until the task finishes executing. Then, only if the return code is non-zero, it'll forward along the output to Parallel.
Parallel doesn't touch the output, it just buffers the output and forwards it along when the task completes as well, which cron will then receive and trigger the email.

So, in a situation where none of the pruners fail, all of the output will be discarded and no email will be triggered.

Little wrappers like `cronic` provide this sort of behavior, but `cronic` in particular also treats any output to `stderr` as if it's worthy of triggering an email. Tarsnapper always writes some output to `stderr`, and we only care about the exit code here, so I did this instead.

Let me know if you have any suggestions to improve the readability here. I agree that it's not very easy to understand, although I hoped my comment would help.

File: README.md (hunk 1 above)
Comment (bdero) [reply README.md]: Yup, done.

File: backup-pruner.yml (hunk 2 above)
Comment (bdero) [reply backup-pruner.yml]: Good call, I've adjusted this. Running all of the pruner scripts in parallel takes between 5 and 6 minutes consistently judging from roughly 10 trials. The backups themselves usually take a bit longer (but still < 15 minutes). I have nothing more than anecdotal evidence to back this up, though. :)

**Comment by bdero** [comment]
> However, the existing design of the backup pruner looks somewhat weird to me. I don't understand why it's necessary to create one script per pruned backup. My preferred approach would be to just have a single script for everything (including the cron job), and to have the configuration in a separate file, so the script is just copied and does not change based on what backups are pruned. (Changing this is of course out of scope for the current story.)

I 100% agree with this. To get this working right away, I chose to work on top of the existing scripts in a way that I trust will be stable, but in the future it would be worth refactoring this into one simple script that pulls in pruner-specific information.
I also don't think being able to prune for individual keys is useful, but even if it did prove useful, this would be an easy thing to support either way.

File: templates/backup-pruner.sh (hunk 3 above)
Comment (smarnach) [reply templates/backup-pruner.sh]: @bdero The bit that I was missing is that GNU parallel (with the parameters you provided) won't have any output other than the output of the child processes.  Maybe you could add a comment pointing this out?  And you could also include that cron sends an email if and only if the cron job writes to stdout or stderr, regardless of the exit code.

Another reason this is hard to read is that it depends on the exact order of the different parsing and subsitution stages in shell command line parsing.  One way of improving this might be to add a flag to all the service-specific scripts to only print their output if they fail.  This would at least make this cron script more readable, but I'm not sure it's worth it overall, given that the whole thing should be refactored the next time we make any major changes.

**Comment by smarnach** [comment]
:+1: Once you've actually pushed the changes you made (I can't see them yet).

- [x] I tested this: This is already running in production, and it works fine.
- [x] I read through the code
- [ ] ~~I checked for accessibility issues~~
- [x] Includes documentation

**Review by smarnach — APPROVED** [review]

**Comment by smarnach** [comment]
@bradenmacdonald Brandon and me agree that the code in this repo could do with some refactoring (see https://github.com/open-craft/deployment-deploy-prune-backups/pull/4#issuecomment-305594475).  I don't think it's urgent (since I hope this thing will just run and do its job now, without needing much hand-holding), so we might not want to add a ticket to the backlog that's just hanging there forever.  Any good suggestion how to remember that comment?

**Comment by bradenmacdonald** [comment]
@smarnach Honestly a ticket is probably the best, or a checklist in an epic. Anything else will likely get forgotten. I will eventually do some work to organize our backlog better and properly deal with "desired but low priority tickets"

File: templates/backup-pruner.sh (hunk 3 above)
Comment (bdero) [reply templates/backup-pruner.sh]: Alright, I've added more specific commentary that should better remind us about these things when we come back to it. :) I'll leave the refactoring for us to do at a later time as you mention.

**Comment by bdero** [comment]
@smarnach I forgot to push the initial changes, sorry about that. All the changes you requested can be seen now (although they're not in a separate commit; I amended them). I added more commentary to the dense couple of lines in the pruner script as well. Could you take another look and let me know if you think it's good? I'm in no rush to merge since we're already running this.

**Comment by smarnach** [comment]
@bdero I had another look at the changes, and all looks good to me.
